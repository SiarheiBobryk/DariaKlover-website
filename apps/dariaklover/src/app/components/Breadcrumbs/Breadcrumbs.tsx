import * as React from 'react';
import { Link as RouterLink, useLocation, LinkProps } from 'react-router-dom';

import BreadcrumbsMui from '@mui/material/Breadcrumbs';
import Link from '@mui/material/Link';
import Typography from '@mui/material/Typography';

import aboutMetaData from '../../pages/About/aboutMetaData';
import articlesMetaData from '../../pages/Articles/articlesMetaData';
import bookingMetaData from '../../pages/Booking/bookingMetaData';
import certificatesMetaData from '../../pages/Certificates/certificatesMetaData';
import faqMetaData from '../../pages/Faq/faqMetaData';
import focusingOnHealthMetaData from '../../pages/FocusingOnHealth/focusingOnHealthMetaData';
import helpMetaData from '../../pages/Help/helpMetaData';
import howToBookMetaData from '../../pages/HowToBook/howToBookMetaData';
import mainMetaData from '../../pages/Main/mainMetaData';
import pricingMetaData from '../../pages/Pricing/pricingMetaData';
import referencesMetaData from '../../pages/References/referencesMetaData';
import { certificatesMap } from '../Certificates';

// SOMEDAY: Refactor breadcrumb names
const breadcrumbNameMapStatic: { [key: string]: string } = {
  [mainMetaData.path]: mainMetaData.breadcrumb,
  [aboutMetaData.path]: aboutMetaData.breadcrumb,
  [articlesMetaData.path]: articlesMetaData.breadcrumb,
  [bookingMetaData.path]: bookingMetaData.breadcrumb,
  [certificatesMetaData.path]: certificatesMetaData.breadcrumb,
  [faqMetaData.path]: faqMetaData.breadcrumb,
  [focusingOnHealthMetaData.path]: focusingOnHealthMetaData.breadcrumb,
  [helpMetaData.path]: helpMetaData.breadcrumb,
  [howToBookMetaData.path]: howToBookMetaData.breadcrumb,
  [pricingMetaData.path]: pricingMetaData.breadcrumb,
  [referencesMetaData.path]: referencesMetaData.breadcrumb,
};

interface LinkRouterProps extends LinkProps {
  to: string;
  replace?: boolean;
}

function LinkRouter(props: LinkRouterProps) {
  return (
    <Link
      {...props}
      component={RouterLink}
      sx={{
        textDecoration: 'none',
        '&:hover': {
          textDecoration: 'underline',
        },
      }}
    />
  );
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface BreadcrumbsProps {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Breadcrumbs = React.forwardRef(function Breadcrumbs(props: BreadcrumbsProps, ref: React.Ref<any>) {
  const location = useLocation();
  const pathnames = React.useMemo(
    function memoizePathnames() {
      return location.pathname.split('/').filter(function filterPathnames(x) {
        return x;
      });
    },
    [location],
  );

  const certificatesBreadcrumbNameMap: { [key: string]: string } = React.useMemo(function memoizeCertificatesBread() {
    return Object.keys(certificatesMap).reduce(function reduceCertificatesMap(certificatesBreadcrumbNameMapRaw, key) {
      const { title, url: titleKey = '' } = certificatesMap[key] ?? {};
      return {
        ...certificatesBreadcrumbNameMapRaw,
        [titleKey]: title,
      };
    }, {});
  }, []);
  const breadcrumbNameMap: { [key: string]: string } = {
    ...breadcrumbNameMapStatic,
    ...certificatesBreadcrumbNameMap,
  };

  return (
    <BreadcrumbsMui ref={ref} aria-label="breadcrumb" sx={{ marginY: 3 }} {...props}>
      <LinkRouter color="inherit" to={mainMetaData.path}>
        Главная
      </LinkRouter>
      {pathnames.map(function mapPathname(value, index) {
        const last: boolean = index === pathnames.length - 1;
        const to = `/${pathnames.slice(0, index + 1).join('/')}`;

        if (!breadcrumbNameMap[to]) {
          // Special case: an unsupported URL is used, so we don't render the data for it
          return null;
        }

        return last ? (
          <Typography color="text.primary" key={to}>
            {breadcrumbNameMap[to]}
          </Typography>
        ) : (
          <LinkRouter color="inherit" to={to} key={to}>
            {breadcrumbNameMap[to]}
          </LinkRouter>
        );
      })}
    </BreadcrumbsMui>
  );
});

export default Breadcrumbs;
