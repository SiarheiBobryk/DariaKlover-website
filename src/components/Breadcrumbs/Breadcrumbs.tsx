import * as React from 'react';

import { Link as RouterLink, useLocation, LinkProps } from 'react-router-dom';

import BreadcrumbsMui from '@mui/material/Breadcrumbs';
import Link from '@mui/material/Link';
import Typography from '@mui/material/Typography';

import aboutMetaData from '../../pages/About/aboutMetaData';
import bookingMetaData from '../../pages/Booking/bookingMetaData';
import helpMetaData from '../../pages/Help/helpMetaData';
import howToBookMetaData from '../../pages/HowToBook/howToBookMetaData';
import mainMetaData from '../../pages/Main/mainMetaData';
import questionsMetaData from '../../pages/Questions/questionsMetaData';
import referencesMetaData from '../../pages/References/referencesMetaData';

const breadcrumbNameMap: { [key: string]: string } = {
  [mainMetaData.path]: mainMetaData.heading,
  [aboutMetaData.path]: aboutMetaData.heading,
  [bookingMetaData.path]: bookingMetaData.heading,
  [helpMetaData.path]: helpMetaData.heading,
  [howToBookMetaData.path]: howToBookMetaData.heading,
  [questionsMetaData.path]: questionsMetaData.heading,
  [referencesMetaData.path]: referencesMetaData.heading,
};

interface LinkRouterProps extends LinkProps {
  to: string;
  replace?: boolean;
}

function LinkRouter(props: LinkRouterProps) {
  return (
    <Link
      {...props}
      component={RouterLink}
      sx={{
        textDecoration: 'none',
        '&:hover': {
          textDecoration: 'underline',
        },
      }}
    />
  );
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface BreadcrumbsProps {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Breadcrumbs = React.forwardRef(function Breadcrumbs(props: BreadcrumbsProps, ref: React.Ref<any>) {
  const location = useLocation();
  const pathnames = React.useMemo(
    function memoizePathnames() {
      return location.pathname.split('/').filter(function filterPathnames(x) {
        return x;
      });
    },
    [location],
  );
  return (
    <BreadcrumbsMui ref={ref} aria-label="breadcrumb" sx={{ marginY: 3 }} {...props}>
      <LinkRouter color="inherit" to={mainMetaData.path}>
        Главная
      </LinkRouter>
      {pathnames.map(function mapPathname(value, index) {
        const last: boolean = index === pathnames.length - 1;
        const to = `/${pathnames.slice(0, index + 1).join('/')}`;

        return last ? (
          <Typography color="text.primary" key={to}>
            {breadcrumbNameMap[to]}
          </Typography>
        ) : (
          <LinkRouter color="inherit" to={to} key={to}>
            {breadcrumbNameMap[to]}
          </LinkRouter>
        );
      })}
    </BreadcrumbsMui>
  );
});

export default Breadcrumbs;
