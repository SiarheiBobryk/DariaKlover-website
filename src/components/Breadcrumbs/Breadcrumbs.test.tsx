import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { axe, toHaveNoViolations } from 'jest-axe';
import { render, RenderResult, screen } from '@testing-library/react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

import Breadcrumbs from './Breadcrumbs';

expect.extend(toHaveNoViolations);

describe('<Breadcrumbs />', () => {
  it('should match a11y standards without props', async () => {
    const { container }: RenderResult = render(
      <BrowserRouter>
        <Breadcrumbs />
      </BrowserRouter>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should forward ref', () => {
    const ref = React.createRef();
    render(
      <BrowserRouter>
        <Breadcrumbs ref={ref} data-testid="Breadcrumbs" />
      </BrowserRouter>,
    );
    expect(screen.getByTestId('Breadcrumbs')).toBe(ref.current);
  });

  it('should render correctly', () => {
    const breadcrumbs: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
      .create(
        <BrowserRouter>
          <Breadcrumbs />
        </BrowserRouter>,
      )
      .toJSON();
    expect(breadcrumbs).toMatchSnapshot();
  });
});
