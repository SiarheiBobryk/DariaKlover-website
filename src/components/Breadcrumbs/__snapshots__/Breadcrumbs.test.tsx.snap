// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should render correctly 1`] = `
<nav
  aria-label="breadcrumb"
  className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1obnsz6-MuiTypography-root-MuiBreadcrumbs-root"
>
  <ol
    className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
  >
    <li
      className="MuiBreadcrumbs-li"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-r3hr3s-MuiTypography-root-MuiLink-root"
        href="/"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        Главная
      </a>
    </li>
  </ol>
</nav>
`;
